File: app/src/main/java/com/trungnguyen/android/houston123/ui/login/LoginViewModel.java
package com.trungnguyen.android.houston123.ui.login;

import android.arch.lifecycle.MutableLiveData;
import android.content.Context;
import android.support.annotation.NonNull;
import android.widget.Toast;

import com.trungnguyen.android.houston123.anotation.OnClick;
import com.trungnguyen.android.houston123.base.BaseViewModel;
import com.trungnguyen.android.houston123.repository.login.AuthenticateRepository;
import com.trungnguyen.android.houston123.repository.login.AuthenticateStore;
import com.trungnguyen.android.houston123.rx.DefaultSubscriber;
import com.trungnguyen.android.houston123.rx.SchedulerHelper;
import com.trungnguyen.android.houston123.util.AppUtils;

import javax.inject.Inject;

import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import timber.log.Timber;

public class LoginViewModel extends BaseViewModel<ILoginView> {

    private Context mContext;
    public LoginModel mLoginModel;

    private boolean mLoginState = false;

    private AuthenticateStore.Repository mAuthRepository;

    @NonNull
    private MutableLiveData isLoggedIn = new MutableLiveData<Boolean>();

    @Inject
    public LoginViewModel(Context context,
                          AuthenticateRepository authenticateRepository) {

        super(context);
        isLoggedIn.setValue(mLoginState);
        this.mContext = context;
        this.mAuthRepository = authenticateRepository;
        mLoginModel = new LoginModel();
    }

    @NonNull
    public MutableLiveData<Boolean> getIsLoggedIn() {
        Disposable disposable = mAuthRepository.getLoginState()
                .compose(SchedulerHelper.applySchedulers())
                .subscribe(isLoggedIn::setValue, Timber::d);
        mSubscription.add(disposable);
        return isLoggedIn;
    }

    @OnClick
    public void onLoginButtonClick() {
        if (!AppUtils.isNetworkAvailable(mContext) && mView != null) {
            mView.showNetworkDialog();
        }
        String userName = mLoginModel.getUserName();
        String password = mLoginModel.getPassword();

        Disposable subscription = mAuthRepository.callLoginApi(userName, password)
                .compose(SchedulerHelper.applySchedulers())
                .doOnSubscribe(disposable -> {
                    if (mView != null) {
                        mView.showLoading();
                    }
                })
                .doOnTerminate(() -> {
                    if (mView != null) {
                        mView.hideLoading();
                    }
                })
                .subscribe(authenticateResponse -> {
                    if (mView != null) {
                        mView.onAuthSuccess(userName);
                    }
                }, throwable -> {
                    if (mView != null) {
                        mView.onAuthSuccess(userName);
                    }
                });

        mSubscription.add(subscription);
    }


    @OnClick
    public void onRegisterClick() {
        Toast.makeText(mContext, "onRegisterClick", Toast.LENGTH_SHORT).show();
    }

    public void putAuthInfoToLocal(boolean loginState, String accessToken) {
        Disposable disposable = mAuthRepository.putAuthInfoLocal(loginState, accessToken)
                .subscribeOn(Schedulers.io())
                .subscribeWith(new DefaultSubscriber<>());
        mSubscription.add(disposable);
    }
}

File: app/src/main/java/com/trungnguyen/android/houston123/ui/login/LoginViewModel.kt
package com.trungnguyen.android.houston123.ui.login

import android.content.Context
import android.text.TextUtils
import com.trungnguyen.android.houston123.anotation.OnClick
import com.trungnguyen.android.houston123.base.BaseViewModel
import com.trungnguyen.android.houston123.data.LoginInfoModel
import com.trungnguyen.android.houston123.data.LoginInfoResponse
import com.trungnguyen.android.houston123.injection.Injector
import com.trungnguyen.android.houston123.repository.login.AuthenticateRepository
import com.trungnguyen.android.houston123.repository.login.AuthenticateStore
import com.trungnguyen.android.houston123.rx.SchedulerHelper
import com.trungnguyen.android.houston123.ui.main.MainActivity
import com.trungnguyen.android.houston123.util.AppUtils
import com.trungnguyen.android.houston123.util.BundleBuilder
import com.trungnguyen.android.houston123.util.SingleLiveEvent
import timber.log.Timber
import javax.inject.Inject

class LoginViewModel
@Inject
constructor(private val mContext: Context, authenticateRepository: AuthenticateRepository) : BaseViewModel<ILoginView>(mContext) {

    var mLoginModel: LoginModel

    private val mAuthRepository: AuthenticateStore.Repository

    val isLoggedIn = SingleLiveEvent<Boolean>()

    val liveUserToken = SingleLiveEvent<String>()

    init {
        this.mAuthRepository = authenticateRepository
        mLoginModel = LoginModel()
    }

    fun getLoginStatus() {
        val disposable = mAuthRepository.loginState
                .compose(SchedulerHelper.applySchedulersLoadingAction({ showLoading() }, { hideLoading() }))
                .subscribe({
                    initGlobalUserCallback(it.convertToModel())
                    handleAuthenticateSuccess("quanly")
//                    isLoggedIn.setValue(true)
                }, {
//                    isLoggedIn.value = false
                    Timber.d(it)
                })
        mSubscription.add(disposable)
    }

    @OnClick
    fun onLoginButtonClick() {
        if (!AppUtils.isNetworkAvailable(mContext) && mView != null) {
            mView!!.showNetworkDialog()
        }
        val userName = mLoginModel.userName
        val password = mLoginModel.password

        if (TextUtils.isEmpty(userName) || TextUtils.isEmpty(password)) {
            return
        }

        val subscription = mAuthRepository.callLoginApi(userName, password)
                .compose(SchedulerHelper.applySchedulers())
                .filter { t: LoginInfoResponse -> !TextUtils.isEmpty("quanly") }
                .doOnNext { response: LoginInfoResponse -> Timber.d("[Auth] User access token [%s]", "quanly") }
                .doOnSubscribe { mView?.showLoadingDialog() }
                .doOnTerminate { mView?.hideLoadingDialog() }
                .subscribe({
                    initGlobalUserCallback(it.convertToModel())
                    handleAuthenticateSuccess("quanly")
                }, {
                    mView?.onAuthFailed()
                })

        mSubscription.add(subscription)

    }


    private fun handleAuthenticateSuccess(permission: String) {
        val clz = MainActivity::class.java
        val bundle = BundleBuilder()
        mView?.onAuthSuccess(permission, clz, bundle.build())
    }

    private fun initGlobalUserCallback(model: LoginInfoModel?) {
        if (model == null) {
            return
        }
        Injector.getInstance().createUserScope(model)
    }
}
